core-java-io/src/test/java/com/baeldung/filechannel/FileChannelUnitTest.java

```diff
@@ -0,0 +1,178 @@
+        String expected_value = "Hello world";
+        String file = "src/test/resources/test_read.in";
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        FileInputStream fin = new FileInputStream(file);
+        FileChannel channel = fin.getChannel();
+
+        int bufferSize = 1024;
+        if (bufferSize > channel.size()) {
+            bufferSize = (int) channel.size();
+        }
+        ByteBuffer buff = ByteBuffer.allocate(bufferSize);
+
+        while (channel.read(buff) > 0) {
+            if (buff.hasArray()) {
+                out.write(buff.array(), 0, buff.position());
+                buff.clear();
+            }
+        }
+
+        assertEquals(expected_value, new String(out.toByteArray(), StandardCharsets.UTF_8));
+        out.close();
+        channel.close();
+        fin.close();
+    }
+
+    @Test
+    public void whenWriteWithFileChannelUsingRandomAccessFile_thenCorrect() throws IOException {
+        String input = "Hello world";
+        String file = "src/test/resources/test_write_using_filechannel.txt";
+
+        RandomAccessFile writer = new RandomAccessFile(file, "rw");
+        FileChannel channel = writer.getChannel();
+
+        ByteBuffer buff = ByteBuffer.wrap(input.getBytes());
+        while (buff.hasRemaining()) {
+            channel.write(buff);
+        }
+        channel.close();
+        writer.close();
+
+        // verify
+        RandomAccessFile reader = new RandomAccessFile(file, "r");
+        assertEquals(input, reader.readLine());
+        reader.close();
+    }
+
+    @Test
+    public void whenWriteWithFileChannelUsingFileOutputStream_thenCorrect() throws IOException {
+        String input = "Hello world";
+        String file = "src/test/resources/test_write_using_filechannel.txt";
+
+        FileOutputStream fout = new FileOutputStream(file);
+        FileChannel channel = fout.getChannel();
+
+        ByteBuffer buff = ByteBuffer.wrap(input.getBytes());
+        while (buff.hasRemaining()) {
+            channel.write(buff);
+        }
+
+        channel.close();
+        fout.close();
+
+        // verify
+        RandomAccessFile reader = new RandomAccessFile(file, "r");
+        assertEquals(input, reader.readLine());
+        reader.close();
+    }
+
+    @Test
+    public void givenFile_whenReadWithFileChannelGetPosition_thenCorrect() throws IOException {
+        long expected_value = 11;
+        String file = "src/test/resources/test_read.in";
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        RandomAccessFile reader = new RandomAccessFile(file, "r");
+        FileChannel channel = reader.getChannel();
+
+        int bufferSize = 1024;
+        if (bufferSize > channel.size()) {
+            bufferSize = (int) channel.size();
+        }
+        ByteBuffer buff = ByteBuffer.allocate(bufferSize);
+
+        while (channel.read(buff) > 0) {
+            if (buff.hasArray()) {
```

Comment:
Could this ever be false given that `channel.read(...)` is `>0`?

Reply:
I read the java doc it is recommended to invoke hasArray() before array() method to ensure that buffer has an accessible backing array.
https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#array() 
This is useful when we get the buffer object as input from other class.

However, since in this case, the same program is reading into the buffer, we can be sure that it will always be backed by an accessible array. hence I am removing this `if` check.